README
NAME

JdCrypt::ByteStream

DESCRIPTION

This module provides a subclass of String with the ability to XOR
it against other String-like objects.

RATIONALE

I needed the ability to XOR strings. I could have created a function
(not method) to do this, but that's not the Ruby way - the Ruby way
is to stick it where it belongs, in the object's method list. I
could have written something(quite easily) to put this function
into String itself, but didn't for three reasons:

- Certain other binary operands (I'm thinking mostly of '~' (binary
NOT) here) would be confusing in the almighty class String, due to
their similarity to existing syntax.

- Byte streams are not strings. They may often look like strings,
but they really really are not. In particular, while a byte stream
could have an encoding, it can't have a charset. It's better to
think of JdCrypt::ByteStream as something that can contain data and
can be conveniently expressed as a string.

- Real Strings would have a different use for XOR: "foo"^"foo"
should really return nil (or "000" or zero), and "foo"^"bar" should
return 1 (or "111"), since you should only consider whether the
characters are different, not their binary representation.

USAGE

foo=JdCrypt::ByteStream.new("Foooooo")
bar=foo^"Baaaaaar"

NOTES

Returned values are still of class JdCrypt::ByteStream, and are of
the same length as the longer value. Note that this means that:

  a=JdCrypt::ByteStream.new("a")
  bb=JdCrypt::ByteStream.new("b")
  a^bb^bb

...does not equal "a" but rather "a\000", so this should be used with
caution except where you have equal-size strings in mind. This
module *could* use some form of padding, but it works this way to
be compatible with perl.

You should also note that ruby handles "operators" in right-to-left
fashion (and it could do it in any order at all), so
JdCrypt::ByteStream.new("a")^"bb"^"bb" will not work;
(JdCrypt::ByteStream.new("a")^"bb")^"bb" will work, but looks very
un-ruby-like. My recommendation for multiple XORs is to put them
on separate lines for now.

Performance is an interesting issue: Ruby does NOT want to deal
with strings (or more specifically arrays of characters) as anything
other than text. For the pure-Ruby core: On both MRI (1.8.x) and
YARV (1.9.x), doing 10000 simple XORs via test.rb takes 0.06-0.09
seconds; on rubinius it's 0.6 seconds. If you're using this for
encryption or decryption, this could be an issue. The binary core
is fast enough to be roughly comparable with OpenSSL.

BUGS

This should use duck-typing, but doesn't. This shouldn't affect
real usage, though.

COPYRIGHT

This files in this distribution are copyright 2005-2011 Jim Driscoll
<jim.a.driscoll@gmail.com>; please see the included file COPYING
for details.


Changelog
v1.0.0
Jim Driscoll (4):
      Support uninstalling the PR version
      BREAKING CHANGE: Dropped all the old compat stuff
      Improve performance testing
      Improve advice a little

v0.9.5
Jim Driscoll (20):
      Fix the pure ruby install
      Minor style (whitespace) cleanup
      Bug: with length mismatch, + would end with nulls
      Port byte mode to the pure-ruby core
      Update & standardise doc
      Whitespace cleanups
      Test update: cover + in both modes
      Re-simplify
      Test more cases
      Port byte-at into the C code
      Actually do return in warning mode
      Explain that special case
      Add some flexibility to the compat mode
      Clearer arg names
      Copy-paste reduction
      Test []
      Make "no" more obvious
      Test strict mode
      Add some compat fixes
      Test compat mode

v0.9.4
Jim Driscoll (3):
      Fix compat with modern Ruby API
      Switch to a non-conflicting name
      Routine code quality update (just whitespace)

v0.9.3
Jim Driscoll (10):
      Fix changelog
      Fix another stupid error
      Avoid triggering own warning
      Nominally support two-arg mode@=
      Actually do support the two-arg mode, as it is unchanged
      Support setting here
      Allow String#[] to be called in some cases
      Passing a range (or whatever) is still fine
      Split the test
      Update release tools

v0.92
Jim Driscoll (5):
      Fix changelog
      Fix stupid error
      Warn about 1.8 <=> 1.9 compat
      Release v0.92
v0.91
Jim Driscoll (8):
      Add a bit of performance stuff to the testing
      Rubinius: ignore compiles scripts
      Clean up documentation
      Note about the speed
      Sensibly specify the mode for install
      Reindent the code a bit
      1.9 compat: add an explicit method to get a particular byte
      Release v0.91
v0.9
Jim A. Driscoll:
      Put bytestream.? in the right place

v0.8
Jim (encrypted):
      No CVS please
      Portability fixes
      Add corresponding "+" func
      TODO list
      More to do
      Make making releases easier
      Show old tag; fix changelog version stuff; changelog updates to top
      Merge ../misc_build_stuff
      Tell CVS to ignore GIT
      ByteStream -> Crypt::ByteStream

0.6
	- Trivial change to fix copyright notice
	- Changelog added


test.rb
# frozen_string_literal: true

require "./bytestream"
foo = JdCrypt::ByteStream.new("fooz")

def assert_warn(value)
  puts(if value
    "Ok"
  else
    "** NO **"
  end)
end

puts "Checking integrity..."
assert_warn(foo == "fooz")
puts "Checking + (32b)..."
assert_warn("#{foo}    " == ["868F8F9A"].pack("H*"))
Foo = JdCrypt::ByteStream.new("FOOX")
puts "Checking + (32b, long)..."
assert_warn("#{Foo}        " == "foox    ")
puts "Checking + (8b)..."
assert_warn("#{foo} " == ["866f6f7a"].pack("H*"))
puts "Checking + (8b, long)..."
assert_warn("#{Foo}     " == "foox ")
puts "Checking XOR..."
foo ^= "123"
assert_warn(foo == "W]\\z")
puts "Checking that XOR won't de-pad..."
foo ^= "fooz"
assert_warn(foo == "123\000")

puts "Checking byte-at (assign)"
bar = JdCrypt::ByteStream.new("bar")
bar.byte_at 1, 32
assert_warn(bar == "b r")

puts "Checking byte-at (read)"
assert_warn(bar.byte_at(0) == "b".ord)

puts "Checking []="
bar = JdCrypt::ByteStream.new("bar")
bar[1] = " "
assert_warn(bar == "b r")

puts "Checking []"
assert_warn(bar[0] == "b")

Before = Time.now
TestTimes = 1_000_000
puts "Testing performance: #{TestTimes} simple XORs"
TestTimes.times do |i|
  foo ^= i.to_s
end
After = Time.now
puts "Took #{After - Before}s"


bytestream.rb
# frozen_string_literal: true

class JdCrypt
  # A subclass of String with a single purpose: to provide the ^ (XOR) operator,
  # for encryption purposes.
  class ByteStream < String
    # Returned values are still of class JdCrypt::ByteStream, and are of the
    # same length as the longer value. Note that this means that:
    # a=JdCrypt::ByteStream.new("a")
    # bb=JdCrypt::ByteStream.new("b")
    # a^bb^bb
    #
    # ...does not equal "a" but rather "a\000", so this should be used with
    # caution except where you have equal-size strings in mind.
    def ^(other)
      max_length = [length, other.length].max
      my_bytes = unpack("C*")
      other_bytes = other.unpack("C*")
      out_bytes = (0..max_length - 1).map do |i|
        (my_bytes[i] || 0) ^ (other_bytes[i] || 0)
      end
      self.class.new(out_bytes.pack("C*"))
    end

    def plus_bytes(other)
      my_bytes = unpack("C*")
      other_bytes = other.unpack("C*")
      max_length = [my_bytes.length, other_bytes.length].max
      overflow = 0
      out_bytes = (0..max_length - 1).map do |i|
        accumulator = (my_bytes[i] || 0) + (other_bytes[i] || 0) + overflow
        overflow = (accumulator & 0x100) >> 8
        accumulator & 0xff
      end
      self.class.new(out_bytes.pack("C*"))
    end

    def plus_words(other)
      my_dwords = unpack("N*")
      other_dwords = other.unpack("N*")
      max_length = [my_dwords.length, other_dwords.length].max
      out_dwords = (0..max_length - 1).map do |i|
        ((my_dwords[i] || 0) + (other_dwords[i] || 0)) & 0xffffffff
      end
      self.class.new(out_dwords.pack("N*"))
    end

    # Binary add with a "String"-like object, return an object of the same class as
    # self. This effectively is little-endian addition in 8-bit or 32-bit style,
    # always carrying to the right.
    #
    # Returned values are of the same length as the longer value. Note that this means
    # that overflows are _dropped_ if both values are the same length.
    def +(other)
      if (other.length % 4).zero? && (length % 4).zero?
        plus_words(other)
      else
        plus_bytes(other)
      end
    end

    Use_getbyte = "".respond_to?(:getbyte)

    def byte_at(position, new_value = nil)
      if new_value
        self[position, 1] = [new_value].pack("C")
      elsif Use_getbyte
        getbyte(position)
      else
        slice(position)
      end
    end

    private :plus_words, :plus_bytes
  end
end


bytestream.c
#include "ruby.h"
#ifdef HAVE_STDINT_H
#include <stdint.h>
#else
#include <machine/types.h>
#endif

VALUE cSelf;

static uint32_t * _binary_add_32(uint32_t *string1, uint32_t length1,
    uint32_t *string2, uint32_t length2);

/*
 * Binary add with a "String"-like object, return an object of the same
 * class as self. This effectively is little-endian addition in 8-bit or 32-bit
 * style, always carrying to the right.
 *
 * Returned values are of the same length as the longer value. Note that this
 * means that overflows are _dropped_ if both values are the same length.
 *
 */

static VALUE bs_binary_add(VALUE self, VALUE in_string) {
    char *self_p;
    char *in_string_p;
    char *out_p;
    long small_length, large_length;
    char i_am_smaller;
    int i;
    long self_len, in_len;
    VALUE string_out;

    in_string_p = RSTRING_PTR(in_string);
    self_p = RSTRING_PTR(self);
    self_len = RSTRING_LEN(self);
    in_len = RSTRING_LEN(in_string);

    /*
     * Find out which is longer (it doesn't matter if neither is)
     */
    if(self_len < in_len) {
        i_am_smaller = 1;
        small_length = self_len;
        large_length = in_len;
    } else {
        i_am_smaller = 0;
        small_length = in_len;
        large_length = self_len;
    }

    if(self_len % 4 == 0 && in_len % 4 == 0) {
        out_p = (char *)_binary_add_32((uint32_t *)self_p,
            self_len / 4, (uint32_t *)in_string_p, in_len / 4);
    } else {
        out_p = (char *)malloc(sizeof(char)* large_length);

        /*
         * For each byte that exists in both strings, add the result into the
         * (start of the) output c-string
         */
        char overflow = 0;
        short accumulator;
        for(i = 0; i < small_length; i++) {
            accumulator = self_p[i] + in_string_p[i] + overflow;
            overflow = (accumulator >= 256) ? 1 : 0;
            out_p[i] = accumulator % 256;
        }
        /*
         * Domino the tail (the bit after the shorter string would end)
         * of the longer string. Note that this is rather inefficient
         * for cases where the overflow is absorbed early.
         */
        if(small_length != large_length) {
            char *larger_string = i_am_smaller ? in_string_p : self_p;
            for(i = small_length; i < large_length; i++) {
                accumulator = larger_string[i] + overflow;
                overflow = (accumulator >= 256) ? 1 : 0;
                out_p[i] = accumulator % 256;
            }
        }
    }

    /*
     * Create a new "String" and massage it into the local class type.
     */
    string_out = rb_str_new(out_p, large_length);
    free(out_p);
    return rb_class_new_instance(1, (VALUE *)&string_out, rb_obj_class(self));
}
static uint32_t * _binary_add_32(uint32_t *string1, uint32_t length1,
    uint32_t *string2, uint32_t length2
) {
    uint32_t *out_p;
    long small_length, large_length;
    char i_am_smaller;
    int i;

    if(length1 < length2) {
        i_am_smaller = 1;
        small_length = length1;
        large_length = length2;
    } else {
        i_am_smaller = 0;
        small_length = length2;
        large_length = length1;
    }

    out_p = (uint32_t *)malloc(sizeof(uint32_t)* large_length);

    uint32_t overflow = 0;
    uint64_t accumulator;
    for(i = 0; i < small_length; i++) {
        accumulator = ntohl(string1[i]) + ntohl(string2[i]) + overflow;
        overflow = (accumulator >= 1ll << 32) ? 1 : 0;
        out_p[i] = htonl(accumulator % (1ll << 32));
    }

    /*
     * Domino the tail (the bit after the shorter string would end)
     * of the longer string. Note that this is rather inefficient
     * for cases where the overflow is absorbed early.
     */
    if(small_length != large_length) {
        uint32_t *larger_string = i_am_smaller ? string2 : string1;
        for(i = small_length; i < large_length; i++) {
            accumulator = ntohl(larger_string[i]) + overflow;
            overflow = (accumulator >= (1ll << 32)) ? 1 : 0;
            out_p[i] = htonl(accumulator % (1ll << 32));
        }
    }

    return out_p;
}

static uint32_t * _binary_xor_32(uint32_t *string1, uint32_t length1,
    uint32_t *string2, uint32_t length2
) {
    uint32_t *out_p;
    long small_length, large_length;
    char i_am_smaller;
    int i;

    /*
     * Find out which is longer (it doesn't matter if neither is)
     */
    if(length1 < length2) {
        i_am_smaller = 1;
        small_length = length1;
        large_length = length2;
    } else {
        i_am_smaller = 0;
        small_length = length2;
        large_length = length1;
    }

    out_p = (uint32_t *)malloc(sizeof(uint32_t)* large_length);

    /*
     * Copy the tail (the bit after the shorter string would end)
     * verbatim from the longer string.
     *
     * But only if there actually is a longer string.
     */

    if(small_length != large_length) {
        memcpy(out_p + small_length,
            (i_am_smaller ? string2 : string1) + small_length,
            (large_length - small_length) * 4);
    }

    /*
     * For each byte that exists in both strings, XOR the result into the
     * (start of the) output c-string
     */
    for(i = 0; i < small_length; i++) {
        out_p[i] = string1[i] ^ string2[i];
    }
    return out_p;
}

/*
 * Binary XOR with a "String"-like object, return an object of the same
 * class as self.
 *
 * Returned values are of the same length as the longer value. Note that this
 * means that:
 *
 * a=JdCrypt::ByteStream.new("a")
 * bb=JdCrypt::ByteStream.new("bb")
 * a^bb^bb
 *
 * ...does not equal "a" but rather "a\000", so this should be used with caution
 * except where you have equal-size strings in mind.
 */

static VALUE bs_binary_xor(VALUE self, VALUE in_string) {
    char *self_p;
    char *in_string_p;
    char *out_p;
    long small_length, large_length;
    char i_am_smaller;
    int i;
    VALUE string_out;
    long self_len, in_len;

    in_string_p = RSTRING_PTR(in_string);
    self_p = RSTRING_PTR(self);

    self_len = RSTRING_LEN(self);
    in_len = RSTRING_LEN(in_string);

    /*
     * Find out which is longer (it doesn't matter if neither is)
     */
    if(self_len < in_len) {
        i_am_smaller = 1;
        small_length = self_len;
        large_length = in_len;
    } else {
        i_am_smaller = 0;
        small_length = in_len;
        large_length = self_len;
    }
    if(self_len % 4 == 0 && in_len % 4 == 0) {
        out_p = (char *)_binary_xor_32((uint32_t *)self_p, self_len / 4,
            (uint32_t *)in_string_p, in_len / 4);
    } else {
        out_p = (char *)malloc(sizeof(char)* large_length);

        /*
         * Copy the tail (the bit after the shorter string would end)
         * verbatim from the longer string.
         *
         * But only if there actually is a longer string.
         */

        if(small_length != large_length) {
            memcpy(out_p + small_length,
                (i_am_smaller ? in_string_p : self_p) + small_length,
                large_length - small_length);
        }

        /*
         * For each byte that exists in both strings, XOR the result into the
         * (start of the) output c-string
         */
        for(i = 0; i < small_length; i++) {
            out_p[i] = self_p[i] ^ in_string_p[i];
        }
    }

    /*
     * Create a new "String" and massage it into the local class type.
     */
    string_out = rb_str_new(out_p, large_length);
    free(out_p);
    return rb_class_new_instance(1, (VALUE *)&string_out, rb_obj_class(self));
}

/*
 * Almost exactly the same as String#to_str. Returns self as a "String".
 */
static VALUE bs_to_str(VALUE self) {
    return rb_str_new(RSTRING_PTR(self), RSTRING_LEN(self));
}

/*
 *
 */

static VALUE bs_byte_at(int argc, VALUE args[], VALUE self) {
    char *self_p;

    if (argc > 2 || argc < 1) {
        rb_raise(rb_eArgError, "wrong number of arguments, only 1 or 2 supported");
    }
    long pos = NUM2LONG(args[0]);
    if(pos > RSTRING_LEN(self) - 1) {
        return Qnil;
    } else {
        self_p = RSTRING_PTR(self);
        if(argc == 2 && !NIL_P(args[1])) {
            self_p[pos] = NUM2INT(args[1]);
        }
        return INT2FIX(self_p[pos]);
    }
}

ID eqeq;
VALUE compatMode;
char compatMode_1_8;

static VALUE bs_compatMode() {
    return compatMode;
}

static VALUE bs_compatMode_assign(VALUE self, VALUE cm) {
    if(strncmp(StringValueCStr(cm), "1.8", 4) == 0) {
        compatMode_1_8 = 1;
    } else {
        compatMode_1_8 = 0;
    }
    compatMode = cm;
    return compatMode;
}

VALUE strictCompat = Qfalse;

static VALUE bs_strictCompat() {
    return strictCompat;
}

static VALUE bs_strictCompat_assign(VALUE self, VALUE sc) {
    strictCompat = sc;
    return strictCompat;
}

static VALUE bs_array_offset(int argc, VALUE argv[], VALUE self) {
    char *self_p;
    VALUE offsetOrRange;

    if (argc > 2 || argc < 1) {
        rb_raise(rb_eArgError, "wrong number of arguments, only 1 or 2 supported");
    }
    if(argc == 2) {
        return rb_call_super(argc, argv);
    } else if(compatMode_1_8) {
        offsetOrRange = argv[0];
        if(TYPE(offsetOrRange) != T_FIXNUM) {
            return rb_call_super(argc, argv);
        } else if(strictCompat == Qtrue) {
            rb_raise(rb_eRuntimeError, "Ambiguous, you must use #byte_at instead");
        } else {
            rb_warn("Ambiguous usage of [], please use #byte_at");
            return rb_call_super(argc, argv);
        }
    } else {
        return rb_call_super(argc, argv);
    }
}

/*
 * A subclass of String with a single purpose: to provide the ^ (XOR) operator,
 * for encryption purposes.
*/
void Init_bytestream() {
    VALUE cJdCrypt=rb_define_class("JdCrypt", rb_cObject);
    VALUE cSelf=rb_define_class_under(cJdCrypt, "ByteStream", rb_cString);
    rb_define_singleton_method(cSelf, "compatMode", bs_compatMode, 0);
    rb_define_singleton_method(cSelf, "compatMode=", bs_compatMode_assign, 1);
    rb_define_singleton_method(cSelf, "strictCompat", bs_strictCompat, 0);
    rb_define_singleton_method(cSelf, "strictCompat=", bs_strictCompat_assign, 1);
    rb_define_singleton_method(cSelf, "strict_mode=", bs_strictCompat_assign, 1);
    rb_define_method(cSelf, "^", bs_binary_xor, 1);
    rb_define_method(cSelf, "+", bs_binary_add, 1);
    rb_define_method(cSelf, "to_str", bs_to_str, 0);
    rb_define_method(cSelf, "byte_at", bs_byte_at, -1);
    rb_define_method(cSelf, "[]", bs_array_offset, -1);

    eqeq = rb_to_id(rb_str_new2("=="));
    compatMode = rb_str_new2("1.8");
    compatMode_1_8 = 1;
}
